File: internal/scanner/scanner.go
package scanner

import (
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"time"

	"github.com/google/uuid"
	"github.com/rophy/av-scanner/internal/cache"
	"github.com/rophy/av-scanner/internal/config"
	"github.com/rophy/av-scanner/internal/drivers"
	"github.com/rophy/av-scanner/internal/metrics"
)

type ScanResponse struct {
	FileID        string              `json:"fileId"`
	Status        drivers.ScanStatus  `json:"status"`
	Engine        config.EngineType   `json:"engine"`
	Signature     string              `json:"signature,omitempty"`
	ScanResult    *drivers.ScanResult `json:"scanResult,omitempty"`
	TotalDuration int64               `json:"totalDuration"`
}

type Scanner struct {
	drivers        map[config.EngineType]drivers.Driver
	activeEngine   config.EngineType
	config         *config.Config
	logger         *slog.Logger
	detectionCache *cache.DetectionCache
}

func New(cfg *config.Config, logger *slog.Logger) *Scanner {
	// Create shared detection cache
	detectionCache := cache.NewDetectionCache(cache.DefaultTTL)

	s := &Scanner{
		drivers:        make(map[config.EngineType]drivers.Driver),
		activeEngine:   cfg.ActiveEngine,
		config:         cfg,
		logger:         logger,
		detectionCache: detectionCache,
	}

	// Initialize drivers with shared cache
	s.drivers[config.EngineClamAV] = drivers.NewClamAVDriver(
		cfg.Drivers[config.EngineClamAV],
		logger,
		detectionCache,
	)
	s.drivers[config.EngineTrendMicro] = drivers.NewTrendMicroDriver(
		cfg.Drivers[config.EngineTrendMicro],
		logger,
		detectionCache,
	)
	s.drivers[config.EngineMock] = drivers.NewMockDriver(
		config.DriverConfig{Engine: config.EngineMock},
	)

	return s
}

// Start starts all driver background watchers
func (s *Scanner) Start() error {
	for engine, driver := range s.drivers {
		if err := driver.Start(); err != nil {
			s.logger.Error("Failed to start driver", "engine", engine, "error", err)
		}
	}
	return nil
}

// Stop stops all driver background watchers
func (s *Scanner) Stop() {
	for _, driver := range s.drivers {
		driver.Stop()
	}
	s.detectionCache.Stop()
}

func (s *Scanner) Scan(filePath, fileID, originalName string, size int64) (*ScanResponse, error) {
	startTime := time.Now()
	driver := s.drivers[s.activeEngine]

	s.logger.Info("Starting scan",
		"fileId", fileID,
		"engine", driver.Engine(),
		"originalName", originalName,
		"size", size,
	)

	absPath, _ := filepath.Abs(filePath)

	// 1. Run manual scan
	result, err := driver.ManualScan(filePath)

	var finalStatus drivers.ScanStatus
	var signature string

	if err == nil && (result.Status == drivers.StatusClean || result.Status == drivers.StatusInfected) {
		// Manual scan completed successfully - use its result
		finalStatus = result.Status
		signature = result.Signature
	} else {
		// 2. Manual scan failed (file missing = RTS quarantined it)
		// Wait for RTS cache with timeout proportional to file size
		driverCfg := driver.Config()
		s.logger.Debug("Manual scan failed, waiting for RTS cache", "error", err, "fileId", fileID)
		retryDelay := 20 * time.Millisecond
		baseDelay := time.Duration(driverCfg.RTSCacheBaseDelay) * time.Millisecond
		delayPerMB := time.Duration(driverCfg.RTSCacheDelayPerMB) * time.Millisecond
		maxWait := baseDelay + time.Duration(size/1024/1024)*delayPerMB
		waited := time.Duration(0)
		for waited < maxWait {
			if cached, found := s.detectionCache.Get(absPath); found && cached.Status == "infected" {
				s.logger.Info("File detected by RTS",
					"fileId", fileID,
					"signature", cached.Signature,
					"waitedMs", waited.Milliseconds(),
				)
				finalStatus = drivers.StatusInfected
				signature = cached.Signature
				break
			}
			time.Sleep(retryDelay)
			waited += retryDelay
		}
		// If still not found in cache, it's an error
		if finalStatus == "" {
			return nil, fmt.Errorf("scan failed: file not accessible and no RTS detection found")
		}
	}

	// 3. Clean up file (may already be removed by RTS)
	s.deleteFile(filePath, fileID)

	response := &ScanResponse{
		FileID:        fileID,
		Status:        finalStatus,
		Engine:        driver.Engine(),
		Signature:     signature,
		ScanResult:    result,
		TotalDuration: time.Since(startTime).Milliseconds(),
	}

	s.logger.Info("Scan completed",
		"fileId", fileID,
		"status", response.Status,
		"duration", response.TotalDuration,
	)

	// Record metrics
	metrics.RecordScan(string(driver.Engine()), string(response.Status))

	return response, nil
}

func (s *Scanner) deleteFile(filePath, fileID string) error {
	if err := os.Remove(filePath); err != nil {
		if os.IsNotExist(err) {
			s.logger.Debug("File already removed (likely by RTS quarantine)",
				"fileId", fileID,
				"filePath", filePath,
			)
			return nil
		}
		return err
	}
	s.logger.Debug("Deleted scanned file", "fileId", fileID, "filePath", filePath)
	return nil
}

func (s *Scanner) CheckHealth() []*drivers.EngineHealth {
	var results []*drivers.EngineHealth
	for _, driver := range s.drivers {
		health, _ := driver.CheckHealth()
		results = append(results, health)
	}
	return results
}

func (s *Scanner) GetActiveEngineHealth() (*drivers.EngineHealth, error) {
	return s.drivers[s.activeEngine].CheckHealth()
}

func (s *Scanner) GetEngineInfo() []drivers.EngineInfo {
	var results []drivers.EngineInfo
	for _, driver := range s.drivers {
		results = append(results, driver.GetInfo())
	}
	return results
}

func (s *Scanner) ActiveEngine() config.EngineType {
	return s.activeEngine
}

func (s *Scanner) GenerateFileID() string {
	return uuid.New().String()
}

func (s *Scanner) GetUploadPath(fileID, originalName string) string {
	ext := filepath.Ext(originalName)
	return filepath.Join(s.config.UploadDir, fileID+ext)
}

File: src/services/scanner.ts
import { unlink } from 'fs/promises';
import path from 'path';
import { v4 as uuidv4 } from 'uuid';
import { AntivirusDriver } from '../drivers';
import { DriverFactory } from './driver-factory';
import {
  AppConfig,
  EngineHealth,
  EngineInfo,
  ScanRequest,
  ScanResponse,
  ScanStatus,
  UnifiedResult,
} from '../types';
import { Logger } from '../utils/logger';

export interface ScanOptions {
  skipManualScan?: boolean;
  rtsTimeout?: number;
}

export class ScannerService {
  private driverFactory: DriverFactory;
  private config: AppConfig;
  private logger: Logger;

  constructor(driverFactory: DriverFactory, config: AppConfig, logger: Logger) {
    this.driverFactory = driverFactory;
    this.config = config;
    this.logger = logger;
  }

  async scan(request: ScanRequest, options: ScanOptions = {}): Promise<ScanResponse> {
    const startTime = Date.now();
    const driver = this.driverFactory.getActiveDriver(this.config);

    this.logger.info('Starting scan', {
      fileId: request.fileId,
      engine: driver.engine,
      originalName: request.originalName,
      size: request.size,
    });

    let rtsResult: UnifiedResult | undefined;
    let manualResult: UnifiedResult | undefined;
    let finalStatus: ScanStatus = 'clean';
    let signature: string | undefined;

    try {
      // Phase 1: Wait for RTS to complete
      if (driver.rtsEnabled) {
        this.logger.debug('Waiting for RTS scan', { fileId: request.fileId });
        rtsResult = await driver.rtsWatch(request.filePath, {
          timeout: options.rtsTimeout || this.config.drivers[driver.engine]?.timeout || 60000,
          pollInterval: 100,
        });

        if (rtsResult.status === 'infected') {
          finalStatus = 'infected';
          signature = rtsResult.signature;
          this.logger.warn('RTS detected infection', {
            fileId: request.fileId,
            signature,
          });
        } else if (rtsResult.status === 'error') {
          this.logger.warn('RTS scan error', { fileId: request.fileId, raw: rtsResult.raw });
        }
      }

      // Phase 2: Manual scan (if RTS was clean and not skipped)
      if (finalStatus === 'clean' && !options.skipManualScan) {
        this.logger.debug('Running manual scan', { fileId: request.fileId });
        manualResult = await driver.manualScan(request.filePath);

        if (manualResult.status === 'infected') {
          finalStatus = 'infected';
          signature = manualResult.signature;
          this.logger.warn('Manual scan detected infection', {
            fileId: request.fileId,
            signature,
          });
        } else if (manualResult.status === 'error') {
          finalStatus = 'error';
          this.logger.error('Manual scan error', { fileId: request.fileId, raw: manualResult.raw });
        }
      }
    } finally {
      // Always delete the file after scanning - security requirement
      await this.deleteFile(request.filePath, request.fileId);
    }

    const response: ScanResponse = {
      fileId: request.fileId,
      status: finalStatus,
      engine: driver.engine,
      signature,
      rtsResult,
      manualResult,
      totalDuration: Date.now() - startTime,
    };

    this.logger.info('Scan completed', {
      fileId: request.fileId,
      status: finalStatus,
      duration: response.totalDuration,
    });

    return response;
  }

  private async deleteFile(filePath: string, fileId: string): Promise<void> {
    try {
      await unlink(filePath);
      this.logger.debug('Deleted scanned file', { fileId, filePath });
    } catch (error) {
      // File may have been quarantined/deleted by RTS
      this.logger.debug('File already removed (likely by RTS quarantine)', {
        fileId,
        filePath,
        error: error instanceof Error ? error.message : 'Unknown error',
      });
    }
  }

  async checkHealth(): Promise<EngineHealth[]> {
    const drivers = this.driverFactory.getAllDrivers();
    const healthChecks = await Promise.all(
      drivers.map((driver) => driver.checkHealth())
    );
    return healthChecks;
  }

  async getActiveEngineHealth(): Promise<EngineHealth> {
    const driver = this.driverFactory.getActiveDriver(this.config);
    return driver.checkHealth();
  }

  getEngineInfo(): EngineInfo[] {
    return this.driverFactory.getAllDrivers().map((driver) => driver.getInfo());
  }

  getActiveEngine(): string {
    return this.config.activeEngine;
  }

  generateFileId(): string {
    return uuidv4();
  }

  getUploadPath(fileId: string, originalName: string): string {
    const ext = path.extname(originalName);
    return path.join(this.config.uploadDir, `${fileId}${ext}`);
  }
}
